composer/utils/checkpoint.py

```diff
@@ -217,17 +217,33 @@ def read_data(self, plan: LoadPlan, planner: LoadPlanner):
         first_replica = self.device_mesh is None or self.device_mesh.ndim == 1 or (
             self.device_mesh.ndim >= 2 and self.device_mesh.get_local_rank(mesh_dim=0) == 0)
 
-        # 1. Download to the destination all files this rank needs if on first replica
+        # 1. Collect the relative paths to download for all ranks for deduplication
+        relative_file_paths = set()
+        for plan_item in plan.items:
+            relative_file_paths.add(self.storage_data[plan_item.storage_index].relative_path)
+        all_file_paths = dist.all_gather_object(relative_file_paths)
+        log.debug(f'Files required for all ranks: {all_file_paths}.')
+
+        # 2. Download to the destination all files this rank needs if on first replica
         if first_replica:
             log.debug(f'Rank {dist.get_global_rank()} starting to download files.')
+
+            # Get the lowest rank in the current node
+            local_rank_0 = dist.get_global_rank() // dist.get_local_world_size() * dist.get_local_world_size()
```

Comment:
```suggestion
            local_rank_0 = dist.get_global_rank() - dist.get_local_rank()
```
mildly cleaner -- my original suggestion was worse. This is useful for variable sized local worlds (not sure why you'd want to do this)